File: test/jvm/com/cattailsw/nanidroid/GhostSwitchingCharacterizationTest.java
package com.cattailsw.nanidroid;

import static org.junit.Assert.assertEquals;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Hashtable;
import java.util.List;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

/**
 * Characterizes the deterministic SScriptRunner ghost-handoff protocol without
 * filesystem discovery, Activity lifecycle, native engines, or view rebinding.
 */
public final class GhostSwitchingCharacterizationTest {
    @Rule
    public final HostAndroidStubRule androidStubs = new HostAndroidStubRule();
    private static final String TRANSITION_SCRIPT = "\\_qSwitching\\e";

    private final Trace trace = new Trace();
    private SScriptRunner runner;
    private RecordingGhost currentGhost;

    @Before
    public void setUp() {
        runner = SScriptRunner.getInstance(null);
        runner.setPresentationRenderer(new TraceRenderer(trace));
        resetRunnerWithPublicApi();
        trace.clear();
    }

    @After
    public void tearDown() {
        resetRunnerWithPublicApi();
    }

    @Test
    public void requiredMigrationInvariant_outgoingScriptRendersBeforeSingleHandoffCallback() {
        RecordingGhost outgoing = new RecordingGhost(
                "outgoing",
                "Old Ghost Metadata",
                "Old Sakura Display",
                1,
                TRANSITION_SCRIPT,
                trace);
        setGhost(outgoing);
        runner.setCallback(new RecordingStatusCallback(trace));

        runner.doGhostChanging("Next Sakura", "manual", "/ghosts/next");

        assertEquals(
                Arrays.asList(
                        "request:outgoing:OnGhostChanging:"
                                + "[Next Sakura, manual, null, /ghosts/next]",
                        "render:Switching",
                        "handoff"),
                trace.events());
    }

    @Test
    public void requiredMigrationInvariant_returningReplacementReceivesChangedFromOutgoingName() {
        RecordingGhost outgoing = new RecordingGhost(
                "outgoing",
                "Old Ghost Metadata",
                "Old Sakura Display",
                1,
                TRANSITION_SCRIPT,
                trace);
        RecordingGhost replacement = new RecordingGhost(
                "replacement",
                "New Ghost Metadata",
                "New Sakura Display",
                2,
                null,
                trace);

        // Prove setup cleanup does not depend on another test having cleared
        // the process singleton's named ghost.
        runner.setGhost(new RecordingGhost(
                "foreign",
                "Foreign Ghost Metadata",
                "Foreign Sakura Display",
                2,
                null,
                trace));
        resetRunnerWithPublicApi();
        trace.clear();

        setGhost(outgoing);
        assertEquals(new ArrayList<String>(), trace.events());
        runner.setCallback(new RecordingStatusCallback(trace));

        runner.doGhostChanging("Next Sakura", "manual", "/ghosts/next");

        setGhost(replacement);

        assertEquals(
                Arrays.asList(
                        "request:outgoing:OnGhostChanging:"
                                + "[Next Sakura, manual, null, /ghosts/next]",
                        "render:Switching",
                        "handoff",
                        "request:replacement:OnGhostChanged:"
                                + "[Old Ghost Metadata, null]"),
                trace.events());
    }

    private void setGhost(RecordingGhost ghost) {
        currentGhost = ghost;
        runner.setGhost(ghost);
    }

    private void resetRunnerWithPublicApi() {
        runner.setNoWaitMode(true);

        // Drain a failed in-flight transition while its callback and inert fake
        // are still installed. This clears changingPending through stop().
        runner.clearMsgQueue();

        runner.setCallback(null);
        runner.setUICallback(null);

        // setGhost(null) dereferences the replacement when the outgoing name is
        // non-null. Suppressing the test fake's name takes the public silent
        // assignment path and avoids coupling this characterization to fields.
        if (currentGhost == null) {
            // A previous or future suite may leave a named ghost in the process
            // singleton. Replacing it with a null-name count-2 fake avoids the
            // production null-replacement dereference without reflection.
            setGhost(new RecordingGhost(
                    "cleanup",
                    null,
                    null,
                    2,
                    null,
                    trace));
        } else {
            currentGhost.suppressOutgoingName();
        }
        runner.setGhost(null);
        currentGhost = null;
    }

    private static final class Trace {
        private final List<String> events = new ArrayList<String>();

        void add(String event) {
            events.add(event);
        }

        void clear() {
            events.clear();
        }

        List<String> events() {
            return new ArrayList<String>(events);
        }
    }

    private static final class RecordingStatusCallback
            implements SScriptRunner.StatusCallback {
        private final Trace trace;

        RecordingStatusCallback(Trace trace) {
            this.trace = trace;
        }

        @Override
        public void stop() {
            // Generic runner-stop notification is outside the handoff oracle.
        }

        @Override
        public void canExit() {
            // Exit handling is outside the handoff oracle.
        }

        @Override
        public void ghostSwitchScriptComplete() {
            trace.add("handoff");
        }
    }

    private static final class RecordingGhost extends Ghost {
        private final String id;
        private String ghostName;
        private String sakuraName;
        private final long createCount;
        private final String transitionScript;
        private final Trace trace;

        RecordingGhost(
                String id,
                String ghostName,
                String sakuraName,
                long createCount,
                String transitionScript,
                Trace trace) {
            super(id);
            this.id = id;
            this.ghostName = ghostName;
            this.sakuraName = sakuraName;
            this.createCount = createCount;
            this.transitionScript = transitionScript;
            this.trace = trace;
        }

        @Override
        protected void loadGhostInfo() {
            // Test-only fake: no descriptors, surfaces, filesystem, or SHIORI engine.
        }

        @Override
        protected void incrementCreateCount() {
            // Test-only fake: create-count values are supplied explicitly.
        }

        @Override
        public String getGhostId() {
            return id;
        }

        @Override
        public String getGhostName() {
            return ghostName;
        }

        void suppressOutgoingName() {
            ghostName = null;
            sakuraName = null;
        }

        @Override
        public long getCreateCount() {
            return createCount;
        }

        @Override
        public String getSakuraName() {
            return sakuraName;
        }

        @Override
        public String getKeroName() {
            return "Kero";
        }

        @Override
        public String getUsername() {
            return "User";
        }

        @Override
        public ShioriResponse doShioriEvent(String event, String[] references) {
            trace.add(
                    "request:" + id + ":" + event + ":"
                            + Arrays.toString(references));
            if ("OnGhostChanging".equals(event) && transitionScript != null) {
                Hashtable<String, String> values = new Hashtable<String, String>();
                values.put("Value", transitionScript);
                return new ShioriResponse("SHIORI/3.0 200 OK", values);
            }
            return new ShioriResponse("SHIORI/3.0 204 No Content");
        }

        @Override
        public void unload() {
            // Ownership and unload ordering are intentionally deferred.
        }
    }

    /** Fail-fast UI-free collaborator for the runner's complete render frame. */
    private static final class TraceRenderer implements GhostPresentationRenderer {
        private final Trace trace;
        private String previousText = "";

        TraceRenderer(Trace trace) {
            this.trace = trace;
        }

        @Override
        public void render(GhostPresentationFrame frame) {
            String value = frame.sakura.text;
            if (!value.equals(previousText) && value.length() > 0) {
                trace.add("render:" + value);
            }
            previousText = value;
        }
    }
}

File: src/test/java/com/cattailsw/nanidroid/GhostSwitchingCharacterizationTest.kt
package com.cattailsw.nanidroid

import java.io.File
import java.util.Arrays
import java.util.concurrent.atomic.AtomicInteger
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout
import org.junit.Assert
import org.junit.Rule
import org.junit.Test

/** Characterizes deterministic runtime-owned ghost handoff without filesystem discovery. */
class GhostSwitchingCharacterizationTest {
    @Rule @JvmField val androidStubs = HostAndroidStubRule()
    @Rule @JvmField val runtimes = RuntimeFixtureRegistry()

    @Test
    fun switchPlaybackOwnsOutgoingResponseBeforeUnload() {
        val trace = Trace()
        val fixture = fixture(trace)
        val outgoing = fixture.requireHandle()
        val targetRoot = File("build/runtime-fixtures/switching/owned-response")
        val operationId = assertIs<RuntimeResult.Success<Long>>(
            fixture.runtime.beginSwitch(outgoing.generation, "owned-response", targetRoot),
        ).value

        fixture.runtime.installTestHooksForTesting(
            GhostRuntimeTestHooks(onOutgoingUnloaded = { trace.add("unload") }),
        ).use {
            Assert.assertTrue(
                fixture.runner.doGhostChanging(
                    operationId,
                    "Owned Response",
                    "manual",
                    targetRoot.path,
                ),
            )
            trace.awaitSize(3)
        }

        Assert.assertEquals(
            listOf(
                "request:outgoing:OnGhostChanging:[Owned Response, manual, null, ${targetRoot.path}]",
                "render:Switching",
                "unload",
            ),
            trace.events(),
        )
    }

    @Test
    fun requiredMigrationInvariant_outgoingScriptRendersBeforeSingleHandoffCallback() {
        val trace = Trace()
        val fixture = fixture(trace)
        val outgoing = fixture.requireHandle()
        val targetRoot = File("build/runtime-fixtures/switching/next")
        val operationId = assertIs<RuntimeResult.Success<Long>>(
            fixture.runtime.beginSwitch(outgoing.generation, "next", targetRoot),
        ).value
        fixture.runtime.installTestHooksForTesting(
            GhostRuntimeTestHooks(onOutgoingUnloaded = { trace.add("handoff") }),
        ).use {
            Assert.assertTrue(
                fixture.runner.doGhostChanging(
                    operationId,
                    "Next Sakura",
                    "manual",
                    "/ghosts/next",
                ),
            )
            trace.awaitSize(3)
        }

        Assert.assertEquals(
            Arrays.asList<String?>(
                "request:outgoing:OnGhostChanging:[Next Sakura, manual, null, /ghosts/next]",
                "render:Switching",
                "handoff",
            ),
            trace.events(),
        )
    }

    @Test
    fun requiredMigrationInvariant_returningReplacementReceivesChangedFromOutgoingName() {
        val trace = Trace()
        val persistence = InMemoryGhostRuntimePersistence().apply {
            activationCounts["replacement"] = 1L
        }
        val fixture = fixture(trace, persistence)
        val outgoing = fixture.requireHandle()
        val targetRoot = File("build/runtime-fixtures/switching/replacement")
        val operationId = assertIs<RuntimeResult.Success<Long>>(
            fixture.runtime.beginSwitch(outgoing.generation, "replacement", targetRoot),
        ).value
        fixture.runtime.installTestHooksForTesting(
            GhostRuntimeTestHooks(onOutgoingUnloaded = { trace.add("handoff") }),
        ).use {
            Assert.assertTrue(
                fixture.runner.doGhostChanging(
                    operationId,
                    "Next Sakura",
                    "manual",
                    "/ghosts/next",
                ),
            )
            val replacement = runBlocking {
                assertIs<RuntimeResult.Success<GhostHandle>>(
                    fixture.runtime.startOrJoin("replacement", targetRoot),
                ).value
            }
            runBlocking {
                assertIs<RuntimeResult.Success<AttachmentReceipt>>(
                    fixture.runtime.attachHost(replacement.generation),
                )
            }
        }

        Assert.assertEquals(
            Arrays.asList<String?>(
                "request:outgoing:OnGhostChanging:[Next Sakura, manual, null, /ghosts/next]",
                "render:Switching",
                "handoff",
                "request:replacement:OnGhostChanged:[Old Ghost Metadata, null]",
            ),
            trace.events(),
        )
    }

    @Test
    fun fatalChangingRequestTerminalizesAnAlreadyJoinedRuntimeSwitch() = runBlocking {
        val failure = com.cattailsw.nanidroid.shiori.ShioriRequestException(
            "outgoing ownership became uncertain",
            ownershipCertain = false,
        )
        val fixture = runtimes.create(
            id = "fatal-outgoing",
            root = File("build/runtime-fixtures/switching/fatal-outgoing"),
            response = { request ->
                if (requestId(request) == "OnGhostChanging") throw failure
                "SHIORI/3.0 204 No Content\r\n\r\n"
            },
        )
        val outgoing = fixture.requireHandle()
        val targetRoot = File("build/runtime-fixtures/switching/fatal-target")
        val operationId = assertIs<RuntimeResult.Success<Long>>(
            fixture.runtime.beginSwitch(outgoing.generation, "fatal-target", targetRoot),
        ).value
        val joined = async(start = CoroutineStart.UNDISPATCHED) {
            fixture.runtime.startOrJoin("fatal-target", targetRoot)
        }

        Assert.assertFalse(
            fixture.runner.doGhostChanging(
                operationId,
                "Fatal Target",
                "manual",
                targetRoot.path,
            ),
        )

        val terminal = withTimeout(1_000L) { joined.await() }
        Assert.assertSame(
            failure,
            assertIs<RuntimeFailure.Fatal>(
                assertIs<RuntimeResult.Failure>(terminal).failure,
            ).cause,
        )
    }

    @Test
    fun poisonedAuthoredPlaybackTerminalizesItsAlreadyJoinedSwitchExactlyOnce() = runBlocking {
        val failure = com.cattailsw.nanidroid.shiori.ShioriRequestException(
            "inline surface request lost native ownership",
            ownershipCertain = false,
        )
        val scheduler = RecordingPlaybackScheduler()
        val fixture = runtimes.create(
            id = "poison-playback-outgoing",
            root = File("build/runtime-fixtures/switching/poison-playback-outgoing"),
            response = { request ->
                when (requestId(request)) {
                    "OnGhostChanging" -> {
                        "SHIORI/3.0 200 OK\r\n" +
                            "Value: \\s[120]\\![open,inputbox,transition]\\e\r\n\r\n"
                    }
                    "OnSurfaceChange" -> throw failure
                    else -> "SHIORI/3.0 204 No Content\r\n\r\n"
                }
            },
            runnerConfiguration = SScriptRunnerConfiguration(
                playbackSchedulerFactory = { scheduler },
            ),
        )
        fixture.runner.setNoWaitMode(true)
        var inputShown = false
        fixture.runner.setUICallback(object : SScriptRunner.UICallback {
            override fun showUserInputBox(id: String) {
                inputShown = id == "transition"
            }

            override fun showUserSelection(textlabel: Array<String>, ids: Array<String>) = Unit
        })
        val outgoing = fixture.requireHandle()
        val targetRoot = File("build/runtime-fixtures/switching/poison-playback-target")
        val operationId = assertIs<RuntimeResult.Success<Long>>(
            fixture.runtime.beginSwitch(
                outgoing.generation,
                "poison-playback-target",
                targetRoot,
            ),
        ).value
        val joinedCompletionCount = AtomicInteger()
        val joined = async(start = CoroutineStart.UNDISPATCHED) {
            fixture.runtime.startOrJoin("poison-playback-target", targetRoot)
        }.also { deferred ->
            deferred.invokeOnCompletion { joinedCompletionCount.incrementAndGet() }
        }

        Assert.assertTrue(
            fixture.runner.doGhostChanging(
                operationId,
                "Poison Playback Target",
                "manual",
                targetRoot.path,
            ),
        )
        Assert.assertTrue(inputShown)
        Assert.assertEquals(GhostRuntimePhase.Poisoned, fixture.runtime.identity().phase)

        fixture.runner.resumeEvt()
        scheduler.runNext()

        val sharedTerminal = withTimeout(1_000L) { joined.await() }
        val directTerminal = fixture.runtime.completeSwitchPlayback(
            outgoing.generation,
            operationId,
        )
        Assert.assertSame(
            failure,
            assertIs<RuntimeFailure.Fatal>(
                assertIs<RuntimeResult.Failure>(sharedTerminal).failure,
            ).cause,
        )
        Assert.assertSame(
            failure,
            assertIs<RuntimeFailure.Fatal>(
                assertIs<RuntimeResult.Failure>(directTerminal).failure,
            ).cause,
        )
        Assert.assertEquals(1, joinedCompletionCount.get())
        Assert.assertEquals(1, fixture.trace.loadCount.get())
        Assert.assertEquals(0, fixture.trace.unloadCount.get())
    }

    @Test
    fun authoredSwitchStaysInteractiveUntilPlaybackCompletionNotifiesTheHost() = runBlocking {
        val scheduler = RecordingPlaybackScheduler()
        val fixture = runtimes.create(
            id = "interactive-outgoing",
            root = File("build/runtime-fixtures/switching/interactive-outgoing"),
            response = { request ->
                if (requestId(request) == "OnGhostChanging") {
                    "SHIORI/3.0 200 OK\r\n" +
                        "Value: \\![open,inputbox,transition]\\e\r\n\r\n"
                } else {
                    "SHIORI/3.0 204 No Content\r\n\r\n"
                }
            },
            runnerConfiguration = SScriptRunnerConfiguration(
                playbackSchedulerFactory = { scheduler },
            ),
        )
        fixture.runner.setNoWaitMode(true)
        var inputShown = false
        fixture.runner.setUICallback(object : SScriptRunner.UICallback {
            override fun showUserInputBox(id: String) {
                inputShown = id == "transition"
            }

            override fun showUserSelection(textlabel: Array<String>, ids: Array<String>) = Unit
        })
        val status = RecordingStatusCallback()
        fixture.runner.setCallback(status)
        val outgoing = fixture.requireHandle()
        val targetRoot = File("build/runtime-fixtures/switching/interactive-target")
        val operationId = assertIs<RuntimeResult.Success<Long>>(
            fixture.runtime.beginSwitch(outgoing.generation, "interactive-target", targetRoot),
        ).value
        val joined = async(start = CoroutineStart.UNDISPATCHED) {
            fixture.runtime.startOrJoin("interactive-target", targetRoot)
        }

        Assert.assertTrue(
            fixture.runner.doGhostChanging(
                operationId,
                "Interactive Target",
                "manual",
                targetRoot.path,
            ),
        )
        Assert.assertTrue(inputShown)
        Assert.assertEquals(GhostRuntimePhase.SwitchPlayback, fixture.runtime.identity().phase)
        Assert.assertEquals(0, status.switchPlaybackCompletions.get())

        val pendingInput = requireNotNull(fixture.runner.dialogueStateSnapshot().pendingInput)
        DialogueDialogBinding { fixture.runner }
            .userInput(pendingInput)
            .onSubmit("transition", "continue")
        scheduler.runNext()

        Assert.assertEquals(1, status.switchPlaybackCompletions.get())
        assertIs<RuntimeResult.Success<GhostHandle>>(withTimeout(1_000L) { joined.await() })
        Unit
    }

    @Test
    fun noScriptSwitchNotifiesHostToEnterReplacementProgress() = runBlocking {
        val fixture = runtimes.create(
            id = "no-script-outgoing",
            root = File("build/runtime-fixtures/switching/no-script-outgoing"),
        )
        val status = RecordingStatusCallback()
        fixture.runner.setCallback(status)
        val outgoing = fixture.requireHandle()
        val targetRoot = File("build/runtime-fixtures/switching/no-script-target")
        val operationId = assertIs<RuntimeResult.Success<Long>>(
            fixture.runtime.beginSwitch(outgoing.generation, "no-script-target", targetRoot),
        ).value
        val joined = async(start = CoroutineStart.UNDISPATCHED) {
            fixture.runtime.startOrJoin("no-script-target", targetRoot)
        }

        Assert.assertTrue(
            fixture.runner.doGhostChanging(
                operationId,
                "No Script Target",
                "manual",
                targetRoot.path,
            ),
        )

        Assert.assertEquals(1, status.switchPlaybackCompletions.get())
        assertIs<RuntimeResult.Success<GhostHandle>>(withTimeout(1_000L) { joined.await() })
        Unit
    }

    private fun fixture(
        trace: Trace,
        persistence: InMemoryGhostRuntimePersistence = InMemoryGhostRuntimePersistence(),
    ): RuntimeFixture = runtimes.create(
        id = "outgoing",
        root = File("build/runtime-fixtures/switching/outgoing"),
        persistence = persistence,
        response = { request ->
            val id = requestId(request)
            if (id == "OnGhostChanging") {
                "SHIORI/3.0 200 OK\r\nValue: $TRANSITION_SCRIPT\r\n\r\n"
            } else {
                "SHIORI/3.0 204 No Content\r\n\r\n"
            }
        },
        preparedFactory = { operationId, ghostId, root ->
            preparedGhost(
                operationId,
                ghostId,
                root,
                name = if (ghostId == "outgoing") "Old Ghost Metadata" else "New Ghost Metadata",
                sakuraName = if (ghostId == "outgoing") "Old Sakura Display" else "New Sakura Display",
                keroName = "Kero",
            )
        },
    ).also { fixture ->
        fixture.trace.requestObserver.set { recorded ->
            val id = requestId(recorded.protocolText)
            if (id == "OnGhostChanging" || id == "OnGhostChanged") {
                trace.add(
                    "request:${recorded.ownerGhostId}:$id:${references(recorded.protocolText)}",
                )
            }
        }
        fixture.runner.setNoWaitMode(true)
        fixture.runner.setPresentationRenderer(TraceRenderer(trace))
    }

    private fun requestId(request: String): String? = request.lineSequence()
        .firstOrNull { it.startsWith("ID: ") }
        ?.removePrefix("ID: ")

    private fun references(request: String): List<String> = request.lineSequence()
        .filter { it.startsWith("Reference") }
        .map { it.substringAfter(": ") }
        .toList()

    private class Trace {
        private val events = java.util.concurrent.CopyOnWriteArrayList<String?>()
        fun add(event: String?) { events += event }
        fun events(): MutableList<String?> = ArrayList(events)
        fun awaitSize(expected: Int) {
            repeat(10_000) {
                if (events.size >= expected) return
                Thread.yield()
            }
            throw AssertionError("trace never reached $expected events: $events")
        }
    }

    private class TraceRenderer(private val trace: Trace) : GhostPresentationRenderer {
        private var previousText = ""
        override fun render(frame: GhostPresentationFrame) {
            val value = frame.sakura.text
            if (value != previousText && value.isNotEmpty()) trace.add("render:$value")
            previousText = value
        }
    }

    private class RecordingPlaybackScheduler : SScriptPlaybackScheduler {
        private val pending = ArrayDeque<() -> Unit>()

        override fun schedule(delayMillis: Long, action: () -> Unit) {
            pending.addLast(action)
        }

        override fun cancelPending() {
            pending.clear()
        }

        fun runNext() = requireNotNull(pending.removeFirstOrNull()).invoke()
    }

    private class RecordingStatusCallback : SScriptRunner.StatusCallback {
        val switchPlaybackCompletions = AtomicInteger()

        override fun stop() = Unit

        override fun canExit() = Unit

        override fun switchPlaybackComplete() {
            switchPlaybackCompletions.incrementAndGet()
        }
    }

    private companion object {
        const val TRANSITION_SCRIPT = "\\_qSwitching\\e"
    }
}
